### ISSUE TITLE
[API Proposal]: Asynchronous DI support

### ISSUE THREAD

**Issue by davidfowl** [dotnet/runtime#65656]
### Background and motivation

There are cases where it is necessary to inject a dependency as a result of an asynchronous operation. This usually comes up when the dependency requires input from some IO operation (like retrieving a secret from a remote store). Today, developers tend to put blocking calls in factories:

```C#
var services = new ServiceCollection();
services.AddSingleton<IRemoteConnectionFactory, RedisConnectionFactory>();
services.AddSingleton<IRemoteConnection>(sp => 
{
    var factory = sp.GetRequiredService<IRemoteConnectionFactory>();
    // NOOOOOO 😢 
    return factory.ConnectAsync().Result;
});
ServiceProvider sp = services.BuildServiceProvider();

IRemoteConnection connection = await sp.GetRequiredServiceAsync<IRemoteConnection>();

public interface IRemoteConnection
{
    Task PublishAsync(string channel, string message);
    Task DisposeAsync();
}

public interface IRemoteConnectionFactory
{
    Task<IRemoteConnection> ConnectAsync();
}
```

The only other viable solution is to move that async operation to method calls, which results in deferring all IO until methods are called (where things can truly be async). This is a non-trivial refactoring that might be impossible depending on the circumstance. The idea here is to provide asynchronous construction support, so that these scenarios can work.

#### Async Resolution Support

```C#
namespace Microsoft.Extensions.DependencyInjection;

public class ServiceDescriptor
{
    public static ServiceDescriptor Describe(Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory, ServiceLifetime serviceLifetime) => null;
}

// These are extension methods that take an async factory
public static class ServiceCollectionExtensions
{
    public static IServiceCollection AddSingleton<TService>(this IServiceCollection services, Func<IServiceProvider, ValueTask<TService>> asyncFactory);

    public static IServiceCollection AddSingleton(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory);

    public static IServiceCollection AddScoped<TService>(this IServiceCollection services, Func<IServiceProvider, ValueTask<TService>> asyncFactory);

    public static IServiceCollection AddScoped(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory);

    public static IServiceCollection AddTransient<TService>(this IServiceCollection services, Func<IServiceProvider, ValueTask<TService>> asyncFactory);

    public static IServiceCollection AddTransient(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory);
}

public static class AsyncServiceProviderExtensions
{
    public static async ValueTask<T> GetRequiredServiceAsync<T>(this IServiceProvider serviceProvider);
    public static async ValueTask<object> GetServiceAsync(this IServiceProvider serviceProvider, Type serviceType);
}
```

#### Async Injection Support

These APIs would use the convention that async resolution is tied to `ValueTask/Task<TServiceType>` and would resolve the service and await the result as part of construction (see the example for more details).

```C#
namespace Microsoft.Extensions.DependencyInjection;

public delegate ValueTask<object> AsyncObjectFactory(IServiceProvider serviceProvider, object?[]? args);
public delegate ValueTask<T> AsyncObjectFactory<T>(IServiceProvider serviceProvider, object?[]? args);

public static class ActivatorUtilities
{
    public static AsyncObjectFactory CreateAsyncFactory(Type type);
    public static AsyncObjectFactory<T> CreateAsyncFactory<T>();
}
```

**NOTE: The generic version could be done using static abstract interface methods and would be more trim friendly.**

### API Usage

```C#
var services = new ServiceCollection();

// This implementation will add `Task<IRemoteConnection>` to the container.
services.AddSingleton<IRemoteConnection>(sp => 
{
    var factory = sp.GetRequiredService<IRemoteConnectionFactory>();
    return factory.ConnectAsync();
});
ServiceProvider sp = services.BuildServiceProvider();

IRemoteConnection connection = await sp.GetRequiredServiceAsync<IRemoteConnection>();

AsyncObjectFactory<B> factory = ActivatorUtilities.CreateAsyncFactory<B>(sp);
B dep = await factory(sp, null);

public class A
{
    private IRemoteConnection _connection;
    private A(IRemoteConnection connection) { _connection = connection }
    // We can't use a constructor here since the container doesn't natively understand async dependencies
    public static A Create(IRemoteConnection connection)
    {
         return new A(connection);
    }
}

public class B
{
    private A _a;
    private B(A a) { _a = a }
    // We can't use a constructor here since the container doesn't natively understand async dependencies
    public static B Create(A a)
    {
         return new B(a);
    }
}

public interface IRemoteConnection
{
    Task PublishAsync(string channel, string message);
    Task DisposeAsync();
}

public interface IRemoteConnectionFactory
{
    Task<IRemoteConnection> ConnectAsync();
}
```

### Risks

- Need to determine what happens when async services are resolved from a sync call to GetService (we should throw).
- ~3rd party DI containers would need to support this~
- There are no async constructors, we need to invent a constructor surrogate for this (see the example).
- ~Implementation complexity (but we can handle this 😄)~. It's a bit easier if we do it as an extension.

**Comment by davidfowl** [comment]
DI council: @alexmg @tillig @pakrym @ENikS @ipjohnson @dadhi @seesharper @jeremydmiller @alistairjevans

**Comment by alexrp** [comment]
FWIW, it feels weird to put `IAsyncServiceProvider` in `Microsoft.Extensions.DependencyInjection` when `IServiceProvider` is in `System`.

**Comment by Joe4evr** [comment]
> * There are no async constructors, we need to invent a constructor surrogate for this.

[Related discussion](https://github.com/dotnet/csharplang/discussions/419), itself linking to [this overarching issue](https://github.com/dotnet/aspnetcore/issues/24142).

**Comment by alistairjevans** [comment]
Ok, so there's a few things to cover off here. Some of them may be specific to Autofac, some of them not. 

We dipped our toe in the async-di water last year to add support for `IAsyncDisposable`.  This actually required relatively minimal changes to the overall Autofac codebase, primarily because disposal happens outside of the actual service resolve pipeline. At the time when we did that, I did consider what might be involved to go full-async, because honestly @davidfowl raising this proposal was just a matter of time.

I'll lead with what you describe as "implementation complexity", just because we should discuss the amount of change needed by us if you want to do async DI properly. Arguably, complexity isn't a superb reason to just not do something, but in this case...

### Async virality

Fundamentally, the entire Autofac resolve pipeline will need to be updated to use `ValueTask<T>` instead of `T`, and async/await as needed. 

We can't only change the resolve path for the async registrations; the viral nature of async prevents it, unless we do sync-over-async when resolving those services.  The ability to await a service resolve would have to be everywhere to actually gain the async benefits.

> Attempting to ignore the viral nature of async by making assumptions in certain places has yielded a [bug](https://github.com/autofac/Autofac/issues/1284) on `IAsyncDisposable` support already.

We also really don't want to maintain a sync _and_ async resolve pipeline, that would be a huge amount of extra code maintenance, because, like I said, everything would have to change in our resolve pipeline.

So that all implies our pipeline becomes async by default, and we add a new `ResolveAsync<TService>()` method on `ILifetimeScope`, that becomes the "normal" resolve entry point. Then the existing sync `Resolve<TService>()` method would invoke `ResolveAsync<TService>()`, and if the returned `ValueTask<TService>` did not complete synchronously we do what? Throw? It's effectively illegal to not await a `ValueTask` you know is completing asynchronously, so not sure what we do there.

```csharp
public class LifetimeScope 
{
    public T Resolve<T>()
    { 
         var task = ResolveAsync<T>();

         if (result.IsCompleted) 
         {
            return task.Result;
         }
         
         // ??
         throw new Exception("?");
    }

    public ValueTask<T> ResolveAsync() 
    {
        // do the resolve
    }
}
```

It's important to note that we cannot know before beginning the resolve whether or not the resolve operation will complete synchronously, so we basically have to "try" and somehow bail after the fact.

This all starts to get pretty complicated.

### Lazy<T>

Users of Autofac often inject `Lazy<T>`, which is based in the sync world.

Specifically, it has a property used to access the underlying service. 

That won't work for async, so I imagine we would need a new `LazyAsync<T>`? Would that live in `System` as `Lazy<T>` does?

Beyond technical problems, this puts an onus on the component to know which type of lazy it should use; this introduces coupling between the service and the dependency, which somewhat defies the point.

### Func<T> Dependencies

Autofac let's you inject invokable factories as `Func<TService>` instances, to be resolved later. I imagine we would need to implement support for generating async versions of those factories.

### Thread Safety

One of the things we've generally asserted in Autofac is that a Resolve happens on a single thread. Because it's possible for users to create new lifetime scopes and generally do very weird stuff in resolve paths, there are locks in place that would need to be replaced, because the resolve would now be able to proceed on a different thread than the one it started on.

### Integrations

Most of our existing active integrations would need updating to understand/support async, which is quite an undertaking in and of itself.

---------

There's almost certainly more challenges here than I've documented; these are just a few. I'm pretty confident that this would be a _huge_ piece of work to implement for Autofac.

Beyond the actual difficulty though...

### Libraries would have to change

These changes could get more complex/confusing for users if an existing library component depends on a service that a user can override. For a random example, let's say a library `Contoso.ConnectLib` has a dependency on Autofac, and uses our `Lazy<T>` support to do lazy instantiation.  (While it may not be best practice for public nuget libraries to take a direct dependency on a IoC container, this happens with internal packages all the time.)

Now, `Contoso.ConnectLib` lets you override some behaviour by registering your own implementation of a service, and a component in the library injects that using `Lazy<IMyService>`.   Now, a user will look at Autofac and think, "cool, I can just register an async delegate for this type". Then they try to use the library and it explodes because you can't do a sync injection of the async type. 

Either the library implementer has to change to support async DI injection, Autofac would have to change to have a compatibility "fallback" of doing sync-over-async, or there will be a lot of users having to apply workarounds like the below to get compatibility with their libraries.

```csharp
// 🤮
builder.Register(ctx => new Lazy<MyService>(() => ctx.ResolveAsync<MyService>().Result));
```

### Performance Overheads

I don't have the numbers on this, but I imagine the runtime team will; what are the overheads of a `ValueTask` backed state machine being added to a method, even if the method completes synchronously? In the Autofac resolve pipeline, it would be sync in a huge percentage of cases.  To do async resolve properly, a lot of methods inside Autofac would get async state machines, and I'm somewhat concerned about the extra CPU time spent doing all resolves, to support a small subset of dependencies doing async setup.

### User Breaking Changes

In 2020, Autofac v6 changed the way that Autofac can be extended by switching to a pipeline approach, and letting users add middleware. Middleware created since then may likely have to change to support users, even if they don't use async factory registration, which I can imagine might annoy our users somewhat.

--------- 

As a parting note, I will add that Autofac is about object construction. That's pretty much it. People have tried to do some pretty crazy things with Autofac beyond that task, up to and including doing their entire app startup using build callbacks and the `IStartable` interface.  Adding async DI like this feels like it could make this worse.

I generally think that injecting a factory type, and then calling async methods on that factory when you need it, to get the appropriate instance, isn't the end of the world, and I'm not sure the trade-offs would be worth it.

**Comment by dadhi** [comment]
I agree with @alistairjevans that the "native" async support will be a very complex task.
So instead, I propose to make this an external extension to solve the specific problem.

If we concentrate only on the async **resolution** (because it is unclear to me how async **injection** suppose to work),
we may use much smaller API extensions of the DI libs to achieve the result.

Here is the working example from the [DryIoc](https://github.com/dotnet/runtime/issues/65656)

```cs
using System;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using NUnit.Framework;

namespace DryIoc.Microsoft.DependencyInjection.Specification.Tests
{
    public static class AsyncExt
    {
        public static IServiceCollection AddSingleton<TService>(this IServiceCollection services, Func<IServiceProvider, Task<TService>> asyncFactory)
        {
            var factoryID = Factory.GetNextID();

            Task<TService> CreateServiceAsync(IServiceProvider sp)
            {
                var dryIoc = sp.GetRequiredService<IResolverContext>();
                var result = dryIoc.SingletonScope.GetOrAddViaFactoryDelegate(factoryID, r => asyncFactory(r), dryIoc);
                return (Task<TService>)result;
            }

            return services.AddSingleton<Func<IServiceProvider, Task<TService>>>(CreateServiceAsync);
        }

        public static Task<TService> GetRequiredServiceAsync<TService>(this IServiceProvider sp) =>
            sp.GetRequiredService<Func<IServiceProvider, Task<TService>>>().Invoke(sp);
    }

    public class AsyncResolutionTestsPOC
    {
        [Test]
        public async Task GetRequiredServiceAsync()
        {
            var services = new ServiceCollection();

            services.AddSingleton<IRemoteConnectionFactory, TestConnectionFactory>();
            services.AddSingleton<IRemoteConnection>(sp =>
            {
                var factory = sp.GetRequiredService<IRemoteConnectionFactory>();
                return factory.ConnectAsync();
            });

            var providerFactory = new DryIocServiceProviderFactory();
            var provider = providerFactory.CreateServiceProvider(providerFactory.CreateBuilder(services));

            var connection1 = await provider.GetRequiredServiceAsync<IRemoteConnection>();
            Assert.IsNotNull(connection1);

            var connection2 = await provider.GetRequiredServiceAsync<IRemoteConnection>();
            Assert.AreSame(connection2, connection1);

            await connection2.PublishAsync("hello", "sailor");
        }

        public interface IRemoteConnection
        {
            Task PublishAsync(string channel, string message);
            Task DisposeAsync();
        }

        public interface IRemoteConnectionFactory
        {
            Task<IRemoteConnection> ConnectAsync();
        }
        class TestConnectionFactory : IRemoteConnectionFactory
        {
            public Task<IRemoteConnection> ConnectAsync() => Task.FromResult<IRemoteConnection>(new TestRemoteConnection());

        }
        class TestRemoteConnection : IRemoteConnection
        {
            public Task DisposeAsync() => Task.CompletedTask;
            public async Task PublishAsync(string channel, string message)
            {
                await Task.Delay(TimeSpan.FromMilliseconds(17));
                Console.WriteLine(channel + "->" + message);
            }
        }
    }
}
```

**Comment by ENikS** [comment]
I went through implementation of async in one of my prototypes for Unity. Ended up with two independent pipelines with rather complex synchronization issues. For example, all thread based lifetimes stopped working in async.

After spending couple of months on this endeavor I decided to scratch support for async completely.

**Comment by davidfowl** [comment]
@alistairjevans Thanks for those thoughts. Those resonate and I understand how complex it would be to support this (as it forks the entire code base).

@dadhi I love this. I've listed it under alternative designs. There are some complexities around:
- What happens when resolution fails. Is the Task cached?
- Synchronizaton needs to be handled in this extension (if multiple people try to resolve the same Task<T>, they should get the same instance).

> (because it is unclear to me how an async injection suppose to work),

I was thinking that we could look at an constructor surrogate. A static factory method on the type (Create or CreateAsync). You could even envision a static abstract interface method here (`ICreateable<T, ...Tn>`)

**Comment by davidfowl** [comment]
I've updated the API proposal. We need to agree on some semantics for how this extension would work. I like the idea that we would map async resolution to `ValueTask/Task<TServiceType>` and we can build other conventions on top of those primitives.

**Comment by dadhi** [comment]
@davidfowl  

> What happens when resolution fails. Is the Task cached?

The demonstrated approach DryIoc will cache the task for the `SingletonScope` or for the  `CurrentScope` lifetime. Moreover `ValueTask` will be boxed to object. 
But this is the current behavior and we may play with that.

> Synchronizaton needs to be handled in this extension (if multiple people try to resolve the same Task, they should get the same instance).

 Not sure. At least `IScope.GetOrAddViaFactoryDelegate` guarantees that the passed task-creation delegate will be called just once.

**Comment by tillig** [comment]
I could be missing something, but it seemed like the convenience of trying to wrap the async factory in the DI container was to abstract away the factory itself so you could "simply inject" the thing that required the async/await.

In the new proposal, it just moves the factory itself out a level, plus it introduces the need for any async resolution to use service location. How is that better than just resolving the factory and using it directly?

**Comment by dazinator** [comment]
I may be staying the onvious here but why not something like this:-

```casharp
var services = new ServiceCollection();
services.AddSingleton<IRemoteConnectionFactory, RedisConnectionFactory>();
services.AddSingleton<Task<IRemoteConnection>>(sp => 
{
    var factory = sp.GetRequiredService<IRemoteConnectionFactory>();
    return factory.ConnectAsync();
});
ServiceProvider sp = services.BuildServiceProvider();

IRemoteConnection connection = await sp.GetRequiredServiceAsync<IRemoteConnection>();

public interface IRemoteConnection
{
    Task PublishAsync(string channel, string message);
    Task DisposeAsync();
}

public interface IRemoteConnectionFactory
{
    Task<IRemoteConnection> ConnectAsync();
}

```

Here `GetRequiredServiceAsync<IRemoteConnection>();` in this scenario is more of a convenience method that resolves `Task<IRemoteConnection>` in the ordinary way.

This seems like a lot less work overall.

Other thoughts:-

Async calls are typically going to be more prone to transient or other exceptions, especially if they involve any network bound stuff.

1. How would exceptions be surfaced?
2. Guidance on resiliency / where would polly policies be placed for example?

I suspect the answers to the above would be that the containers wouldn't help here, and so that stuff would have to be "handled" by the developer in their underlying factory method implementations..

I think an Async factory method for the purposes of DI, is a slightly different beast than an async method for the purposes of fetching a secret from a keystore. If you have the latter it doesn't mean you should place that in the former location without wrapping it with sufficient resiliency or other features and this means you always want a factory method that you can control.

**Comment by mrudat** [comment]
I would expect that a failed async initialisation (e.g. database is down) would have the exception handled precisely the same way as throwing an exception in a constructor is now (e.g. an invalid requested state in the new object).

I'm not sure why it might _want_ to be semantically distinct. It's still a failure to create the object you wanted.

If you require special handling for a transient error, you can do that in the async constructor or a try/catch, as usual.

**Comment by mrudat** [comment]
On the subject of _consuming_ services that might be created asynchronously...

I think that you want to be able to request _both_ `TService` and `Task<TService>` for _all_ services, regardless of how they're created/registered.

If your class doesn't do any async initialisation, you don't need to care. If you do, you can request a Task for everything (regardless of how it was created) and only await it when required.

That means when or if a dependency changes to async construction, your code doesn't need to change to take advantage of it.

For example, if you don't have any async initialisation in the second class, both classes have the same behaviour.

```cs
public class SyncConstructor(TService service) {}

public class AsyncConstructor{
    private AsyncConstructor(TService service) {}

    [AsyncConstructor] // or however we declare an async factory.
    public static async Task<AsyncConstructor> Factory(Task<TService service>) {
        // async stuff here...
        return new AsyncConstructor(await service);
    }
}
```

Of course, it only makes sense to wait for _potentially_ async objects, but you could, for example, start requesting l10n from a slow resource, and it's suddenly an async service.

**Comment by rosebyte** [comment]
**TL;DR** Making asynchronous services support on a par with synchronous services would force every DI container (MEDI included) into substantial rewrites and compromises they may decide to reject because the overwhelming majority of services will always be synchronous and DI sits in hot paths. Moreover, consumers have to be on the new async APIs before libraries can safely expose async services without sync-over-async hacks. The move should therefore be split into steps, starting with a small API surface for the straightforward, application-level cases and adding the most-demanded features iteration by iteration. To handle the differences between containers, I propose to implement a shim that serves as a reasonably performing fallback, so the API works with any container out of the box while container authors who want to offer more can implement the API natively in their container.

# Requirements
Collected from both this issue and the dead-ends I found when working on multiple spikes. I'll keep it brief for now, but if anyone is interested we can discuss any of them deeper.
1. **Services aren't observable until they're ready:** No caller can obtain a reference to an asynchronously constructed service until construction is awaited (this prevents, among other things, the initialization race when a shared service is requested by multiple threads).
2. **Sync-over-async is avoidable:** we need to allow users to fit their async-initialized objects into DI flows without needing to do sync-over-async, which might harm the DI container's sync-first inner mechanisms (which is the case of MEDI's locking mechanism).
3. **Allow containers their own pace**: there are multiple DI containers in .NET, and they may want to set their own pace when it comes to async services. Some containers may have a more async-friendly architecture than others, and their authors may see a higher or lower value in this move; we need to give them a way to position their container where it makes sense to them.
4. **Constructor injection only supports instances ready to be consumed:** Constructors are synchronous and DI should encourage it, otherwise we risk sync-over async in constructors and won't be able to guarantee proper dispose order..
5. **Disposable instances are properly disposed:** both `IDisposable` and `IAsyncDisposable` instances produced by async factories must be tracked by the container and disposed exactly once when their owning scope (or the root, for singletons) is disposed. At the same time, only disposable instances should be tracked, to avoid wasting memory on bookkeeping for everything else.
6. **Lifetimes are properly handled:** singleton means one instance per root, scoped means one instance per scope, transient means one instance per resolution — and in every case the factory runs at most once per such slot. Concurrent resolvers await the same shared completion and observe the same instance.
7. **Captive-dependency rules apply unchanged:** a scoped dependency of a singleton must be rejected for the same reason it is today.
8. **Introspection must be enabled for async services too:** async services will be isolated from sync services to comply with the other requirements, so we will likely need to add async counterparts to `IsService` / `IsKeyedService`.
9. **Resolution must be allocation-light and lock-light:** DI resolution lives on application hot paths (per-request scopes, repeated transient resolves), and we need to consider this for any design we want to propose.

# API Proposal

```csharp
public static class ServiceCollectionServiceExtensions
{
    public static IServiceCollection AddSingleton<T>(this IServiceCollection services, Func<IServiceProvider, ValueTask<T>> asyncFactory) where T : class { throw null; }
    public static IServiceCollection AddSingleton(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddScoped<T>(this IServiceCollection services, Func<IServiceProvider, ValueTask<T>> asyncFactory) where T : class { throw null; }
    public static IServiceCollection AddScoped(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddTransient<T>(this IServiceCollection services, Func<IServiceProvider, ValueTask<T>> asyncFactory) where T : class { throw null; }
    public static IServiceCollection AddTransient(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddKeyedSingleton<T>(this IServiceCollection services, object? serviceKey, Func<IServiceProvider, object?, ValueTask<T>> asyncFactory) where T : class { throw null; }
    public static IServiceCollection AddKeyedSingleton(this IServiceCollection services, Type serviceType, object? serviceKey, Func<IServiceProvider, object?, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddKeyedScoped<T>(this IServiceCollection services, object? serviceKey, Func<IServiceProvider, object?, ValueTask<T>> asyncFactory) where T : class { throw null; }
    public static IServiceCollection AddKeyedScoped(this IServiceCollection services, Type serviceType, object? serviceKey, Func<IServiceProvider, object?, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddKeyedTransient<T>(this IServiceCollection services, object? serviceKey, Func<IServiceProvider, object?, ValueTask<T>> asyncFactory) where T : class { throw null; }
    public static IServiceCollection AddKeyedTransient(this IServiceCollection services, Type serviceType, object? serviceKey, Func<IServiceProvider, object?, ValueTask<object>> asyncFactory) { throw null; }
}

public static class ServiceProviderServiceExtensions
{
    public static ValueTask<T?> GetServiceAsync<T>(this IServiceProvider provider) where T : class { throw null; }
    public static ValueTask<T> GetRequiredServiceAsync<T>(this IServiceProvider provider) where T : class { throw null; }
}

public static class ServiceProviderKeyedServiceExtensions
{
    public static ValueTask<T?> GetKeyedServiceAsync<T>(this IServiceProvider provider, object? serviceKey) where T : class { throw null; }
    public static ValueTask<T> GetRequiredKeyedServiceAsync<T>(this IServiceProvider provider, object? serviceKey) where T : class { throw null; }
}

public class AsyncServiceDescriptor : ServiceDescriptor
{
    public Type UnderlyingServiceType { get { throw null; } }
    public Func<IServiceProvider, ValueTask<object>>? AsyncImplementationFactory { get { throw null; } }
    public Func<IServiceProvider, object?, ValueTask<object>>? KeyedAsyncImplementationFactory { get { throw null; } }
    public AsyncServiceDescriptor(ServiceLifetime lifetime, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { }
    public AsyncServiceDescriptor(ServiceLifetime lifetime, Type serviceType, object? serviceKey, Func<IServiceProvider, object?, ValueTask<object>> asyncFactory) { }
    public static AsyncServiceDescriptor Describe<T>(ServiceLifetime lifetime, Func<IServiceProvider, ValueTask<T>> asyncFactory) where T : class { throw null; }
    public static AsyncServiceDescriptor DescribeKeyed<T>(ServiceLifetime lifetime, object? serviceKey, Func<IServiceProvider, object?, ValueTask<T>> asyncFactory) where T : class { throw null; }
}

public partial interface IAsyncServiceProvider
{
    ValueTask<object?> GetServiceAsync(Type serviceType);
    ValueTask<object> GetRequiredServiceAsync(Type serviceType);
}

public interface IKeyedAsyncServiceProvider : IAsyncServiceProvider
{
    ValueTask<object?> GetKeyedServiceAsync(Type serviceType, object? serviceKey);
    ValueTask<object> GetRequiredKeyedServiceAsync(Type serviceType, object? serviceKey);
}

public interface IAsyncServiceProviderIsService
{
    bool IsAsyncService(Type serviceType);
}

public interface IKeyedAsyncServiceProviderIsService : IAsyncServiceProviderIsService
{
    bool IsAsyncKeyedService(Type serviceType, object? serviceKey);
}
```

The two *IsService interfaces are optional and not covered by the shim because that would limit services in either implementing them or keep using the shim. The behaviour is that by default these interfaces aren't implemented and if any container wants to provide this introspection for async registrations, it needs to implement them or let the current sync *IsService methods returning true on asynchronously resolved services. MEDI will implement them to return either wrapped or sync services is present in the container, and then change it if the shim would be ever abandoned.
The existing IsService(typeof(T)) continues to answer truthfully about what the underlying sync container knows; for a purely async registration that answer is "no", because the sync API isn't able to provide the service. The same logic applies to GetServiceAsync\<T> and its required and keyed counterparts: if no async registration is found, the extension methods fall back to synchronous resolution, and only then do they return null or throw. The general idea is that async resolution can handle (and therefore includes) synchronous resolution, but the reverse isn't true.

# Sample

```csharp
// Imagine a remote cache whose connection has to be established asynchronously,
// and a query engine that depends on it.

public interface IRemoteCache : IAsyncDisposable
{
    ValueTask<string?> GetAsync(string key);
}

public sealed class RemoteCache : IRemoteCache
{
    private RemoteCache() { }
    public static async ValueTask<RemoteCache> ConnectAsync(string endpoint)
    {
        var cache = new RemoteCache();
        await cache.HandshakeAsync(endpoint);
        return cache;
    }
    private Task HandshakeAsync(string endpoint) => Task.CompletedTask;
    public ValueTask<string?> GetAsync(string key) => default;
    public ValueTask DisposeAsync() => default;
}

public sealed class QueryEngine(IRemoteCache cache)
{
    public ValueTask<string?> RunAsync(string q) => cache.GetAsync(q);
}

// Registration. Note that QueryEngine depends on an async-resolved service,
// so QueryEngine itself must also be registered through an async factory,
// and it pulls its dependency through GetRequiredServiceAsync.

var services = new ServiceCollection();

services.AddSingleton<IRemoteCache>(async sp =>
{
    var endpoint = sp.GetRequiredService<IConfiguration>()["cache:endpoint"]!;
    return await RemoteCache.ConnectAsync(endpoint);
});

services.AddSingleton<QueryEngine>(async sp =>
    new QueryEngine(await sp.GetRequiredServiceAsync<IRemoteCache>()));

// Keyed variant.
services.AddKeyedSingleton<IRemoteCache>("primary", async (sp, _) =>
    await RemoteCache.ConnectAsync("primary.example.com"));

services.AddKeyedSingleton<IRemoteCache>("secondary", async (sp, _) =>
    await RemoteCache.ConnectAsync("secondary.example.com"));

// Non-generic (Type-based) variant, useful for bridges and reflection-driven code.
services.AddSingleton(typeof(IRemoteCache), async sp =>
    await RemoteCache.ConnectAsync("default.example.com"));

IServiceProvider sp = services.BuildServiceProvider();

// Consumption. Async resolution is a strict superset of sync resolution
// at the call site, so a hosted service can simply await whatever it needs.

public sealed class Startup(IServiceProvider provider) : IHostedService
{
    public async Task StartAsync(CancellationToken ct)
    {
        QueryEngine engine = await provider.GetRequiredServiceAsync<QueryEngine>();
        IRemoteCache primary = await provider.GetRequiredKeyedServiceAsync<IRemoteCache>("primary");
        // ... engine and primary are fully constructed by the time we get here.
    }

    public Task StopAsync(CancellationToken ct) => Task.CompletedTask;
}

// Introspection.

var probe = sp.GetRequiredService<IServiceProviderIsService>();
bool hasCache = probe is IAsyncServiceProviderIsService isservice && isservice.IsAsyncService(typeof(IRemoteCache);        // true

var keyedProbe = sp.GetRequiredService<IServiceProviderIsKeyedService>();
bool hasPrimary = probe is IAsyncKeyedServiceProviderIsService iskeyedservice && iskeyedservice.IsKeyedServiceAsync<IRemoteCache>("primary"); // true

// A foreign container can short-circuit the shim by implementing IAsyncServiceProvider;
// the extension methods above will detect it and dispatch directly.
```

# Shim
The shim turns every async registration into a synchronous registration of an opaque wrapper type, so the underlying container can do everything it already does (scope walks, lifetime bookkeeping, disposal tracking, concurrency control) without knowing anything about asynchrony. The await lives at the call site, inside the consumer-side extension methods. When an async service is registered, it's stored in the service collection as an AsyncServiceDescriptor (which inherits from ServiceDescriptor) and keeps all the information a container needs to handle the rest of the story on its own; if the container chooses not to, it will just see another ServiceDescriptor for the Async<T> wrapper type. This type is only meant to handle the asynchronous flow for resolution, while remaining subject to whatever behaviour the container applies to services in general. If the factory returns synchronously, the wrapper simply stores the value; otherwise it stores the task and orchestrates any further requesters so that they all wait until the service is ready. For singletons and scoped services the wrapper is always disposable, because there is no risk in capturing a no-op disposable in these scenarios. If the service is transient, the wrapper is itself a disposable scoped service that also holds a list of the transient instances (if they are disposable) and disposes them as part of the scope's disposal. If the factory throws synchronously (before the first await), the exception surfaces to the caller and the wrapper does not cache the failed task; the next resolution re-invokes the factory. If the factory faults after yielding, the wrapper's Task\<T> is observed as faulted by every awaiter and stays faulted for the slot's lifetime, so subsequent resolutions see the same exception. This matches the well-understood semantics of Lazy\<Task\<T>> and re-entering a partially-completed factory while earlier awaiters' continuations may still be running would violate the "one factory run per slot" guarantee. Container authors who want native behaviour have two seams, each independently sufficient:
- At build time. Walk the collection, look for AsyncServiceDescriptor entries, and process them through your own async-aware factory pipeline using the public AsyncImplementationFactory / KeyedAsyncImplementationFactory. Do not install the shim's wrapper registration alongside.
- At resolution time. Implement IAsyncServiceProvider (and IKeyedAsyncServiceProvider). The consumer-side extension methods detect this and dispatch to your implementation directly, bypassing the wrapper machinery entirely. You can ignore the wrapper registrations or treat them as your own bookkeeping; the consumer never sees them.
A container that adopts neither still serves async registrations correctly through the shim, with the wrapper-shaped behaviour described above. A container that adopts both can decide which one wins.

# Limitations
The wrapper should serve the vast majority of use cases we expect for asynchronous services well, but there are a few limitations that follow from the way it's implemented. Some of these are deliberate (they uphold the requirements above), while others are pragmatic and could be lifted later if there is demand.
1. **No open generics**: Open-generic registrations (`AddSingleton(typeof(IFoo<>), typeof(Foo<>))`) have no async counterpart in this proposal. Resolving an open generic against an async factory would mean materialising the factory shape per closed type at resolution time, which the shim cannot do without per-type code generation. Until we see a real scenario, this is left as future work.
2. **IEnumerable\<T> will not contain asynchronously constructed Ts**: Today MEDI resolves IEnumerable\<T> by walking every descriptor registered for T and synchronously building each entry. An async-registered service is held in the container as an Async\<T> wrapper, not as a T, so IEnumerable\<T> will not see it. A potential IAsyncEnumerable\<T> resolution path could fix this in the future, but it is out of scope here.
3. **Async services cannot be constructor-injected**: This is a deliberate consequence to prevent users awaiting them in synchronous constructors or return instances depending on unawaited members. Async dependencies are pulled with an explicit `await sp.GetRequiredServiceAsync<TDep>()` inside another async factory body. Async is viral by design, because the alternative is unsafe.
4. **Only generic resolution methods on the consumer side, no Type-based counterparts**: The interface methods on IAsyncServiceProvider and IKeyedAsyncServiceProvider take a Type and return ValueTask<object?>; the generic, type-checked overloads (GetServiceAsync\<T>, GetRequiredServiceAsync\<T>, and so on) live on top of that interface as extension methods. This mirrors the long-standing split between IServiceProvider.GetService(Type) and the generic GetService\<T>() extension, and it lets container authors implement the asynchronous seam once and pick up the generic surface for free.
5. **No value-type services**: Every async overload is constrained where T : class. Value types are awkward in MEDI today (they tend to box, they trip up several optimisations, and the existing test matrix already struggles with them on some platforms), and we don't want to inherit that pain. If a strong scenario appears, the constraint can be relaxed in a follow-up; until then the limitation is explicit at compile time rather than as a runtime surprise.
6. **No CancellationToken**: This is closer to a requirement than a limitation, because the semantics of cancellation on a shared async resolution don't really hold up. Imagine a singleton whose factory is triggered by thread A and then awaited by thread B as well: if B cancels, the work is still running for A; if A cancels, B is happy to keep waiting. There's no single answer to "what does cancellation mean here" that doesn't either lie to one of the callers or break the "one factory run per slot" invariant. Callers who want to bound their own wait can attach a token to the returned ValueTask themselves via task.WaitAsync(ct); factories that want to honour a token can capture one through a closure. Both are honest about scope, which the API parameter would not be.

# Microsoft Extensions DI container
MEDI now relies heavily on synchronous methods in every aspect. Over the years many optimisations have been made, and they often traded execution flexibility for performance, which makes perfect sense when everything stays synchronous. But it also means we would lose a lot of performance if we simply changed all the internal methods to return `ValueTask<object?>` instead of plain `object?`. If we wanted to make asynchronously built services first-class citizens in MEDI, we would therefore need to fork almost the whole current container implementation, which is a lot of work and brings real risks. Besides, synchronous services will still be the vast majority of what MEDI constructs, and the question is whether asynchronous services will turn out to matter to users, or be a niche feature that people want to have available but don't really care about in terms of non-functional requirements.
MEDI will not implement native async resolution in the first iteration. The shim is enough to serve the scenarios in #65656 today, and shipping the shim first lets us learn what async-DI usage actually looks like in the wild before we pay for a native implementation in the default container. Native async resolution in MEDI would mean reworking the existing synchronous resolution pipeline (call-site cache, generic resolution, the locking around singleton construction, disposal tracking) to interleave with awaitable state. That is a non-trivial complexity and performance investment, and we should be confident it pays for itself in measured scenarios before committing to it.

# Alternatives
There are a couple of alternatives I have put aside:
1. **Completely async DI**: This would mean rewriting a lot of the current DI: both the compiled-expression engine and the IL-emission engine would always have to consider whether dependencies are async and wait for them. The worst of it is that we couldn't keep current performance for synchronously built services. Aside from the higher complexity of the code paths, every `object?` return value on the resolution path would have to become `ValueTask<object?>`. ValueTask is a fairly large struct (24 B) and copying it by value through the call chain costs around 3 ns per call, which can amount to roughly 30 ns per request. That doesn't sound like much, but it is a regress that all services would pay, including the entirely synchronous ones (which will always be the vast majority). DI can sit in hot paths where this would matter a lot.
2. **Hybrid DI**: The idea is to identify, per service, whether it needs async resolution, and keep the synchronous path for services where we can guarantee there will be no waits. That preserves most of the current DI, at the cost of writing new paths for asynchronous services and living with multiple execution paths (which tends to produce surprises here and there). Another issue is asynchronous toxicity: because consumers can't be sure that someone won't later change a dependency to an asynchronous implementation, they will tend to put async API counterparts everywhere just to be safe. That isn't critical (for most services it's just one ValueTask jump, comparable to the concurrent-dictionary read we already dislike on IServiceProvider today), but it does spread. This alternative might still work, but I'd recommend we ship the shim first and revisit only if the shim turns out not to be enough and we haven't found a better way during its evaluation period.
3. **Register Task\<T> directly**: The idea is to register Task\<T> for asynchronously built services and rely on the shape of the type to communicate that the result has to be awaited; otherwise it can be used as any other dependency, even in constructors (we'd encourage sync-over-async at construction time in that case). The biggest issue is that Task\<T> inherits Task, which implements IDisposable, and that Dispose throws when the task hasn't been awaited, instead of awaiting the result and disposing it. That can leave produced services undisposed. On top of that, every single resolved instance would be captured and disposed when the owning scope is disposed even when the result isn't disposable, which can lead to memory waste (and effectively leaks) when transient non-disposable services are resolved within a long-lived scope.
4. **Public wrapper**: The idea is to make the wrapper class awaitable and public, which sidesteps the Task-implements-IDisposable problem. The public API would be smaller too: there's simply a wrapper type that can be requested and then awaited (again, sync-over-async at construction time encouraged); we'd pay one extra allocation of the wrapper per resolved instance, which is amortisable for singleton and scoped services but an issue for transient ones. The biggest concern, though, is that the wrapper type would stay forever, calcified right in the centre of the DI API. Even if we later decided that a different API would serve us better (likely direct container support for async services), the wrapper would remain as an alternative way of doing things in perpetuity, becoming yet another regret point in the .NET API surface. With this design there is also no opt-out for containers: we'd be dictating the flow to them too.

**Comment by svick** [comment]
@rosebyte  This sounds good to me overall, just a few comments about the exact API shape:

```c#
public static bool IsServiceAsync<T>(this IServiceProviderIsService probe) where T : class
public static bool IsKeyedServiceAsync<T>(this IServiceProviderIsKeyedService probe, object? serviceKey) where T : class
```

The `Async` suffix generally means that the method returns a `Task` (or a similar type). I think better names for these methods would be `IsAsyncService` and `IsAsyncKeyedService`.

---

Don't we need to have `IsServiceAsync` on some interface (either e.g. `IServiceProviderIsServiceAsync` or directly on `IAsyncServiceProvider`), for containers that don't want to use the shim?

---

The existing `IsService` uses `Type`, the proposed `IsServiceAsync` uses `<T>`, shouldn't those be symmetric? If a generic type is considered better, should we add a generic version of `IsService` (as an extension method)?

**Comment by rosebyte** [comment]
@ENikS, @tillig, @alexmg, @seesharper, @jeremydmiller, @dadhi, @z4kn4fein, @ipjohnson, do you have any comments on the proposal?

**Comment by rosebyte** [comment]
@svick, nice catch. The *IsService interfaces should be there for exactly the reasons you mentioned. Moreover, the generic extension method does not have any synchronous counterpart, so I am removing it. Each container can now choose whether to support introspection for async services.

**Comment by tillig** [comment]
Autofac [has largely chimed in on this already](https://github.com/dotnet/runtime/issues/65656#issuecomment-1046962469) so I'm not going to repeat what was already said, but the key point remains:

> As a parting note, I will add that Autofac is about object construction. That's pretty much it. People have tried to do some pretty crazy things with Autofac beyond that task, up to and including doing their entire app startup using build callbacks and the `IStartable` interface. Adding async DI like this feels like it could make this worse.

I think that if there's some sort of shim that allows for folks to get async behavior without having to change the underlying containers, awesome; but considering DI - _dependency injection_ - is about _object construction_, and there is no such thing (right now) as "async constructors," it seems like adding something like this is going beyond what DI is actually for.

I'll not be happy the day I get an issue for "why is _Autofac_ running slow during _object construction_" when the person is trying to initialize a connection to a SQL Server and a file share during DI. I think there's a balance between "making this easy/removing boilerplate" and "masking potential foot-gun issues." This feels more like the latter than the former. If you want async initialization, inject the factory and run the async method yourself so you explicitly acknowledge the impact.

**Comment by rosebyte** [comment]
@tillig, thank you for the feedback. I read your post about trying to make Autofac async-ready, and I agree that containers may remain synchronous indefinitely. Even for MEDI, it is not certain it would ever make that step to async. Therefore, the design is intentionally made as a shim that enables the most obvious asynchronous use cases almost for free, while also providing hooks if any container chooses to take that step. We can ship just the shim first, with the rest following as appropriate, so the net11.0 API change would look roughly like this:

# API Proposal

```csharp
public static class ServiceCollectionServiceExtensions
{
    public static IServiceCollection AddAsyncSingleton<T>(this IServiceCollection services, Func<IServiceProvider, ValueTask<T>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncSingleton<T>(this IServiceCollection services, object? key, Func<IServiceProvider, ValueTask<T>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncSingleton(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncSingleton(this IServiceCollection services, object? key, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncScoped<T>(this IServiceCollection services, Func<IServiceProvider, ValueTask<T>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncScoped<T>(this IServiceCollection services object? key, Func<IServiceProvider, ValueTask<T>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncScoped(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncScoped(this IServiceCollection services, object? key, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncTransient<T>(this IServiceCollection services, Func<IServiceProvider, ValueTask<T>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncTransient<T>(this IServiceCollection services, object? key, Func<IServiceProvider, ValueTask<T>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncTransient(this IServiceCollection services, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
    public static IServiceCollection AddAsyncTransient(this IServiceCollection services, object? key, Type serviceType, Func<IServiceProvider, ValueTask<object>> asyncFactory) { throw null; }
}

public static class ServiceProviderServiceExtensions
{
    public static ValueTask<T?> GetAsyncService<T>(this IServiceProvider provider) { throw null; }
    public static ValueTask<T?> GetAsyncService<T>(this IServiceProvider provider, object? key) { throw null; }
}
```

# Sample

```csharp
// Imagine a remote cache whose connection has to be established asynchronously,
// and a query engine that depends on it.

public interface IRemoteCache : IAsyncDisposable
{
    ValueTask<string?> GetAsync(string key);
}

public sealed class RemoteCache : IRemoteCache
{
    private RemoteCache() { }
    public static async ValueTask<RemoteCache> ConnectAsync(string endpoint)
    {
        var cache = new RemoteCache();
        await cache.HandshakeAsync(endpoint);
        return cache;
    }
    private Task HandshakeAsync(string endpoint) => Task.CompletedTask;
    public ValueTask<string?> GetAsync(string key) => default;
    public ValueTask DisposeAsync() => default;
}

public sealed class QueryEngine(IRemoteCache cache)
{
    public ValueTask<string?> RunAsync(string q) => cache.GetAsync(q);
}

// Registration. Because IRemoteCache is registered through an async factory,
// it cannot be constructor-injected. QueryEngine depends on it, so QueryEngine
// itself must also be registered as async and pull its dependency through
// GetAsyncService inside the factory body.

var services = new ServiceCollection();

services.AddAsyncSingleton<IRemoteCache>(async sp =>
{
    var endpoint = sp.GetRequiredService<IConfiguration>()["cache:endpoint"]!;
    return await RemoteCache.ConnectAsync(endpoint);
});

services.AddAsyncSingleton<QueryEngine>(async sp =>
    new QueryEngine((await sp.GetAsyncService<IRemoteCache>())!));

// Keyed variant.
services.AddAsyncSingleton<IRemoteCache>("primary", async sp =>
    await RemoteCache.ConnectAsync("primary.example.com"));

services.AddAsyncSingleton<IRemoteCache>("secondary", async sp =>
    await RemoteCache.ConnectAsync("secondary.example.com"));

// Non-generic (Type-based) variant, useful for bridges and reflection-driven code.
services.AddAsyncSingleton(typeof(IRemoteCache), async sp =>
    (object)await RemoteCache.ConnectAsync("default.example.com"));

IServiceProvider sp = services.BuildServiceProvider();

// Consumption. Async resolution at the call site is a strict superset of sync
// resolution: a hosted service can simply await whatever it needs.

public sealed class Startup(IServiceProvider provider) : IHostedService
{
    public async Task StartAsync(CancellationToken ct)
    {
        QueryEngine? engine = await provider.GetAsyncService<QueryEngine>();
        IRemoteCache? primary = await provider.GetAsyncService<IRemoteCache>("primary");
        // ... engine and primary are fully constructed by the time we get here.
    }

    public Task StopAsync(CancellationToken ct) => Task.CompletedTask;
}
```

There would be no work required from Autofac or any other container. We can use net11.0 to collect feedback and observations, then decide what to add for net12.0.

I’ve changed the naming slightly to emphasise that async services are not just another set of services from DI, so it should feel less like a footgun. What do you think?

**Comment by Alxandr** [comment]
The one thing I feel we loose with this (that we kind of get for free with sync services) is the wiring of dependencies. When I do `AddSingleton<MyService>()`, the MEDI will find the constructor to use and construct the dependency graph itself. I think loosing that ability (which I believe this design sacrefices) is a rather high cost. I would really like to see a convention or similar for something like an `public static ValueTask<MyService> CreateAsync()` as the async counterpart to a constructor. This would also enable async counterparts of `ActivatorUtilities` helpers that produce `ObjectFactory`s, so we could have an `AsyncObjectFactory`. The alternative is a lot of boilerplate in every application, which is likely going to result in the feature being largely unused.

**[Edit]**

On an unrelated note - should the factories and invokers (`GetAsyncService`) accept/provide cancellation-tokens? It's pretty odd to have async methods without them these days...

**Comment by tillig** [comment]
If it doesn't require changes to the underlying container, maybe ship it as a separate library that isn't part of the base framework. I'm not sure why it'd need to be owned here. 

It seems like this is going to logically end in one of two not-great spots:

1. People like it but don't understand why the async bits don't inject into constructors. It'll get people doing sync-over-async hacks to "bridge the gap" and filing complaints. Foot-gun.
2. It will "encourage" more service location because the execution of the async part of things is tied to the service provider. Tests become harder to write and dependencies harder to determine. Scenarios like AOT become harder. Foot-gun.

I know this probably comes across as the classic "engineer who always says no" but, as noted, we looked at this already. I feel like it falls into the category of "just because we can doesn't mean we should."

And, again, if this can be shipped as a shim that doesn't touch the ServiceCollection or ServiceProvider... Just ship that as its own library for people to opt in and use if desired.

**Comment by rosebyte** [comment]
@Alxandr, the shim is intended as a first step before deciding whether to make a larger investment into async DI. Auto‑wiring is a dominant part of container implementation, at least for MEDI, so implementing it would effectively mean building an async DI upfront. The cancellation token absence is explained above:
>  the semantics of cancellation on a shared async resolution don't really hold up. Imagine a singleton whose factory is triggered by thread A and then awaited by thread B as well: if B cancels, the work is still running for A; if A cancels, B is happy to keep waiting. There's no single answer to "what does cancellation mean here" that doesn't either lie to one of the callers or break the "one factory run per slot" invariant. Callers who want to bound their own wait can attach a token to the returned ValueTask themselves via task.WaitAsync(ct); factories that want to honour a token can capture one through a closure. Both are honest about scope, which the API parameter would not be.

**Comment by rosebyte** [comment]
@tillig, shipping it in a different package (either in dotnet/runtime or dotnet/extensions) is definitely an option. Shipping it as a third-party package would not help other parts of .NET that are limited to first-party packages, but that is a separate discussion.

1. People already do sync-over-async in constructors, which is one of the motivations to provide a better-defined way to obtain correctly scoped and wired services. Of course, the most determined can take a dependency on IServiceProvider and then resolve the async service in the constructor using sync-over-async, but it would actually be easier for them to continue doing what they do now. So I understand your concern about sync-over-async, but I believe we would, if not improve things, at least not make them worse.
2. I dare say it would encourage service location in much the same way as any service using a registration factory today. The concept of registration factories is not new, and so far it has not led to anything catastrophic. If your point is that async factories only compose through registration factories, so if you have an async dependency you need such a factory for all consumers and their consumers, that is true, but that is how async propagation works. If we do not want to go fully async, we do not have many other options because, as you mentioned, there are no async constructors in the first place.

I am proposing this with several motivations mentioned above, like avoiding users doing sync-over-async in constructors and in synchronous registration factories, or providing a way for engineers to obtain correctly initialised, scoped, and owned services without effectively reimplementing what this feature would do internally over and over again.

**Comment by Alxandr** [comment]
> Imagine a singleton whose factory is triggered by thread A and then awaited by thread B as well: if B cancels, the work is still running for A; if A cancels, B is happy to keep waiting.

Right - but I still think you should have cancellation tokens. I've implemented similar solutions previously, and there are a couple of reasons where it would make a lot of sense:

1. The (async) scope might be cancelled:

(note, written in browser, so probably remembered some method names wrong)

```csharp
var scope = container.CreateAsyncScope();
var serviceTask = scope.ServiceProvider.GetAsyncService<IFoo>();
await scope.DisposeAsync();
```

2. One of the consumers might want to bail fast

The way I've implemented this previously is by doing something like this:

```csharp
// assuming this is how we implement GetAsyncService
public static ValueTask<T> GetAsyncService<T>(this IServiceProvider s, CancellationToken ct)
  => s.GetRequiredService<AsyncWrapper<T>>().GetAsync().WaitAsync(ct);
```

This makes cancellation work for each instansiation - in the weird cases where you have multiple. And the instance is still owned by the scope/container.

**Comment by quixoticaxis** [comment]
> 1. The (async) scope might be cancelled:
> 
> (note, written in browser, so probably remembered some method names wrong)
> 
> ```csharp
> var scope = container.CreateAsyncScope();
> var serviceTask = scope.ServiceProvider.GetAsyncService<IFoo>();
> await scope.DisposeAsync();
> ```

It's a misuse even in synchronous DI. You're trying to bring a dependency out of scope. Btw passing `CancellationToken` **in** won't help here. To detect scope being cleaned, you'd need to pass a token **out**, or use some entirely different form of notification channel.


> 2. One of the consumers might want to bail fast
> 
> The way I've implemented this previously is by doing something like this:
> 
> ```csharp
> // assuming this is how we implement GetAsyncService
> public static ValueTask<T> GetAsyncService<T>(this IServiceProvider s, CancellationToken ct)
>   => s.GetRequiredService<AsyncWrapper<T>>().GetAsync().WaitAsync(ct);
> ```
> 
> This makes cancellation work for each instansiation - in the weird cases where you have multiple. And the instance is still owned by the scope/container.

It raises the questions about the desired behaviors. What if the instance is already created and the last client cancels, should the instance be disposed on the spot?

Side note: I'm sceptical about "one size fits all" async DI: too many nuanced behaviors to standardize for little gains. Even common DI abstractions already cut-off parts of .Net eco system, and bringing async to the table would only broaden the gap.

**Comment by Alxandr** [comment]
> It's a misuse even in synchronous DI. You're trying to bring a dependency out of scope. Btw passing `CancellationToken` **in** won't help here. To detect scope being cleaned, you'd need to pass a token **out**, or use some entirely different form of notification channel.

The cancellation token in this case would be passed to the factory. The goal here is to be able to stop connecting to the cache/database/whatever if the scope is disposed. The `serviceTask` in the (contrived) example above would also be cancelled (if it has not syncrhonously completed).

> It raises the questions about the desired behaviors. What if the instance is already created and the last client cancels, should the instance be disposed on the spot?

The instance is owned by the DI. If construction has started, it will not be stopped by consumers cancelling tokens. Only by disposing of the scope/container. The reason for allowing the passing of a cancellation-token is to be able to exit whatever method called `GetAsyncService` using cooperative cancellation. The actual construction of the service would still continue in the background, but by throwing `OperationCancelledException` from `GetAsyncService` it's likely to unwind the (async) stack enough that the scope too would be disposed in most cases I assume. That wouldn't matter if the service was a singleton though.

**Comment by rosebyte** [comment]
@Alxandr, the reason isn't ergonomic, it's that a `CancellationToken` parameter would have to lie about what it cancels. The factory runs at most once per slot and its result is observed by every joiner. If thread A triggers the factory and thread B joins the wait, B's token can't cancel A's work, and A's token can't release B. We'd ship a parameter that only ever cancels the local await while looking like it cancels the resolution.
 
The one token whose lifetime actually lines up with a resolution is the scope's disposal, because the scope owns the resolution. And we already act on that without a parameter by checking scope disposal before handing the instance back, so an instance produced after the scope is gone is disposed rather than orphaned (moreover, we would need to do this part even if we added the cancellation token as a parameter to prevent leaks when users ignore it). Any caller-supplied token would either be the wrong shape (asymmetric across joiners) or a redundant copy of one we already honour.
 
The remaining angle is "I want the factory itself to bail out mid-await once the scope is going away, to stop wasting resources", even though disposal of whatever it does produce is taken care of regardless. That doesn't need an API change either, it's a few lines of user code:
 
```csharp
internal sealed class ScopeLifetime : CancellationTokenSource
{
    protected override void Dispose(bool disposing)
    {
        if (disposing && !IsCancellationRequested) Cancel();
        base.Dispose(disposing);
    }
}

services.AddScoped<ScopeLifetime>();
services.AddAsyncSingleton<IRemoteCache>(async sp =>
{
    var ct = sp.GetRequiredService<ScopeLifetime>().Token;
    return await RemoteCache.ConnectAsync(endpoint, ct);
});
```

That gives a token whose lifetime is honestly the scope's, opt-in for the factories that care, and zero impact on the API surface or on factories that don't. A caller who instead wants to bound their own local wait can attach a token to the returned `ValueTask` themselves (`task.AsTask().WaitAsync(ct)`, or the equivalent shim on older TFMs). Both are accurate about whose lifetime the token represents, which a method parameter wouldn't be.

**Comment by rosebyte** [comment]
@quixoticaxis, I understand your concerns and share them to the extent that making async initialised services a first class citizen, by which I mean supporting all their use cases, would require a major investment in the DI system and most likely a new DI altogether.
The proposal has a smaller goal: to provide a limited subset that helps users obtain asynchronously initialised services with proper scoping and disposal, and ensures they are only accessible when ready, while avoiding sync over async, which causes issues for synchronous DIs in terms of locking and other internal mechanisms.

**Comment by quixoticaxis** [comment]
> The cancellation token in this case would be passed to the factory. The goal here is to be able to stop connecting to the cache/database/whatever if the scope is disposed. 

The initial point above was about threads A and B, so I assumed we do speak about multiple stacks and outer scope that is not tied to either.
Everything's kind of fine in a "single call stack" scenario, when essentially the stack owns the scope and scope owns the objects, but async enables lots of scenarios when scope is outside and there are multiple independent call stacks.


> The instance is owned by the DI.

"Single call stack" looks fine. Yet again, we get to a point in a more decoupled scenarios when there could be an object that lives in the scope but has never been reachable by user from wherever, because everyone cancelled their DI requests.

My point is: `async` enables/encourages multi-context scenarios with the scope (living far away up the async call stack) that persists over cancellation at the bottom of the stack and thus can potentially become a storage for unreachable stale state. (And a source of hard-to-debug bugs. )

**Comment by quixoticaxis** [comment]
@rosebyte sorry, if I missed it in the pages above, but was the idea of `async` `Lazy`-like considered? Or even DI-based `Lazy`-like in a sense like the current DI-based `Activator`.
The wrapper would 1. work with synchronous DI and 2. would be visible to the user and thus would allow the user to handle it freely.

**Comment by Alxandr** [comment]
> We'd ship a parameter that only ever cancels the local await while looking like it cancels the resolution.

I still think this is worth it for the following scenario:

```csharp
private async Task SomeFunc(CancellationToken ct) 
{
    var svc = await _service.GetAsyncService<MyService>(); // this takes 10 minutes
    await svc.DoSomething(ct);
}
```

If I call `SomeFunc` here with a cancellation-token that times out after 20 seconds for instance, I would be "stuck" in the `GetAsyncService` call. Considering the fact that there might be (bad) networking going on in the construction of the service makes this something that can reasonably happen in production code. I would much rather have it accept a cancellation token and just document the fact that it does not cancel the actual resource acquisition, but just stops waiting for it.

**Comment by rosebyte** [comment]
@quixoticaxis the wrapper does follow AsyncLazy in spirit (cached completion, joiners share it), but doesn't use any public BCL type like `Lazy<Task<T>>` because they aren't disposable in a way we need to guarantee correct disposal of the inner service. Moreover, with any public type, users would be tempted to request it directly in constructor parameters, leading to sync over async in constructors. So we so far use an internal lazy-like fit-tailored type.

**Comment by quixoticaxis** [comment]
> @quixoticaxis the wrapper does follow AsyncLazy in spirit (cached completion, joiners share it), but doesn't use any public BCL type like `Lazy<Task<T>>` because they aren't disposable in a way we need to guarantee correct disposal of the inner service. Moreover, with any public type, users would be tempted to request it directly in constructor parameters, leading to sync over async in constructors. So we so far use an internal lazy-like fit-tailored type.

I was thinking about a new type, something along the lines of:

```csharp
AsyncInitialized<T> : IAsyncDisposable
{
    Task<T> GetValueAsync();
}
```
which would be explicitely requested.
That way the issue of `async` factory delegate moves from DI to consumer's code. Yes, the user would be tempted to block on the call in costructor, where the wrapper is requested, but the user would be tempted to go sync over async in lots of places anyway. I struggle to see the boundary between the current proposal and essentially upgrading DI to be a full asynchronous workflow engine.

**Comment by quixoticaxis** [comment]
@rosebyte, please, bear with me a little as I suppose I've gathered my thoughts on the matter and what is the problem I see.

I guess I fail to see who the user of the API is and how it's going to facilitate better ergonomic usage.

1. Current case.
The consumer designs some class, but does not think too much about whether or not some asynchronous initialization is needed. When he's done and tries to tie everything together with DI, he encounters _compile-time_ error, becuase his factory method does not compile. The developer chooses a quick and dirty fix, adds `.Wait()`, or `.GetAwaiter().GetResult()`, or whatever, and now everything works, but the delegate invoked by DI blocks.
2. Using the proposed API.
The consumer of the API designs some class as in 1, but now he sees a new overload, adds `Create` method and everything compiles. Now the _compile-time_ error becomes a _runtime_ error (because the class can be implicitely instantiated synchronously) at indeterminate points in code and time, some of which are possibly not known to the author of the code nor controlled by him.
3. Using an async initialization wrapper API.
The consumer of the API does the same as in 1, but now the newly found overload does not register `T` in DI, it registers `Wrapper<T>`, `AsyncInitialized<T>`, `LazyAsyncDependecy<T>`, or whatever it's called. Now the _compile-time_ error becomes _runtime_ error, but only through the code the developer introduced himself. He's free to choose how to handle it. He, of cource, can choose to go with sync-over-async instread of re-designing the code, but now it's his responsibility and does not affect DI.

It seems to me I miss some crucial point about the current proposal, because, as far as I can see, the proposed API does not facilitate better coding practices and does not limit potential risks to the newly introduced code.

**Comment by rosebyte** [comment]
@Alxandr the polyfill `vt.AsTask().WaitAsync(ct)` does exactly what the parameter overload would do internally, but keeping it at the call site has three concrete advantages over hiding it behind syntactic sugar:

- It does not signal that the operation is cancellable, because it is not
- We can first check if the call returned synchronously before setting up the wait (potentially allocating a cancellation source for that)
- It will not trigger CA2016, which would otherwise push users to always provide a cancellation token, even though they usually would not benefit from it

**Comment by rosebyte** [comment]
@quixoticaxis, I’m not sure I understand (2) - which `Create` method do you mean? The idea with the proposed API is that users register a factory returning the instance once it is ready. If there is any dependency that needs to be awaited, they do so within the registration factory, including calling a dependency via `provider.GetAsyncService<MyType>()`.

There is no plan to make the wrapper public, so sync-over-async in (3) cannot happen by design.

> I am proposing this with several motivations mentioned above, like avoiding users doing sync-over-async in constructors and in synchronous registration factories, or providing a way for engineers to obtain correctly initialised, scoped, and owned services without effectively reimplementing what this feature would do internally over and over again.

**Comment by quixoticaxis** [comment]
@rosebyte, the original post contains the following example (and the `Create` method I mentioned comes from there):
```
public class A
{
    private IRemoteConnection _connection;
    private A(IRemoteConnection connection) { _connection = connection }
    // We can't use a constructor here since the container doesn't natively understand async dependencies
    public static A Create(IRemoteConnection connection)
    {
         return new A(connection);
    }
}

public class B
{
    private A _a;
    private B(A a) { _a = a }
    // We can't use a constructor here since the container doesn't natively understand async dependencies
    public static B Create(A a)
    {
         return new B(a);
    }
}
```
the original posts also lists a risk:

> Need to determine what happens when async services are resolved from a sync call to GetService (we should throw).

What happens if I inject my asynchronously initialized `IRemoteConnection` into `A`, `A` into `B` and some code I do not even know of requests `B` synchronously? Should it throw?

> I am proposing this with several motivations mentioned above, like avoiding users doing sync-over-async in constructors and in synchronous registration factories, or providing a way for engineers to obtain correctly initialised, scoped, and owned services without effectively reimplementing what this feature would do internally over and over again.

How is injection supposed to work in general? I do not understand it from the original proposal.

**Comment by rosebyte** [comment]
@quixoticaxis, I see, the original post is from 2022, the current proposal is [here](https://github.com/dotnet/runtime/issues/65656#issuecomment-4450227447) and the reduced version [here](https://github.com/dotnet/runtime/issues/65656#issuecomment-4548794667), please take a look.

**Comment by quixoticaxis** [comment]
> [@quixoticaxis](https://github.com/quixoticaxis), I see, the original post is from 2022, the current proposal is [here](https://github.com/dotnet/runtime/issues/65656#issuecomment-4450227447) and the reduced version [here](https://github.com/dotnet/runtime/issues/65656#issuecomment-4548794667), please take a look.

I see, thank you for the links. AFAIU, the reduced version does codify the contract of async factory, so people can use it without introducing their own interfaces, and does only that, right?

How does reduced version play with `IsService` method? The same as the non-reduced version, so synchronous DI knows nothing about the asynchronously initialized types?

**Comment by tillig** [comment]
I've followed this thread for a long time and want to summarize my position clearly so I can step away from the discussion. I mean no disrespect to anyone who's invested time here — there's been thoughtful work, and I appreciate that.

**My core concern remains unchanged**: DI containers are object graph composers. Their job is to construct and wire up objects. Asynchronous initialization - connecting to databases, fetching secrets, warming caches - is application logic that belongs in the application's startup sequence, not inside the container's resolution pipeline.

I acknowledge that people do sync-over-async in factories today. But I don't think the answer to "people are misusing the tool" is "make the tool easier to misuse." The answer is better guidance and patterns - things like `IHostedService`, explicit initialization phases, or injecting factories that return `Task<T>` so the consumer owns the async flow.

The shim proposal is technically clever, and I respect the engineering behind it. But it results in services that can't be constructor-injected, which means consumers must explicitly call `GetAsyncService<T>()` - that's service location with extra steps. If we've lost auto-wiring, the primary benefit of a DI container, what are we actually gaining over injecting a `Func<Task<T>>` or an `IAsyncFactory<T>`?

Finally, if the shim works without container changes, it can ship as a standalone package. Putting it in the framework signals endorsement of the pattern, which I believe will lead to more problems than it solves - particularly as developers start putting slow I/O in resolution paths, making containers unpredictably slow and hard to reason about.

I'll leave it there. I don't think I have anything new to add beyond this, but I wanted to be clear about where I stand. Thanks to everyone for the civil discussion.

**Comment by rosebyte** [comment]
@quixoticaxis, yes, the goal is to make it possible to use DI in this way, primarily to prevent users from doing sync-over-async in constructors.

The reduced version abstracts this so that IsService will not return true for anything users cannot obtain via IServiceProvider.GetService.
